src/sender.rs

```diff
@@ -1,85 +1,499 @@
+pub enum SendError<T> {
+    Closed(T),
+    Existing(T),
+}
+
+impl<T> SendError<T> {
+    /// Get the value that failed to send back.
+    pub fn into_inner(self) -> T {
+        match self {
+            SendError::Closed(t) => t,
+            SendError::Existing(t) => t,
+        }
+    }
+}
+impl<T> Deref for SendError<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        match self {
+            SendError::Closed(t) => t,
+            SendError::Existing(t) => t,
+        }
+    }
+}
+
+/// A unique means of sending a message on a Hatch.
+///
+/// ## Options (all false by default)
+///
+/// * `overwrite` - when true, overwrites any existing value.
+/// * `close_on_send` - when true, closes during the next send
+///   operation that successfully delivers a message.
+/// * `mark_on_drop` - when true, marks the atomic with a sentinel value
+///   for an external process (not provided!) to reclaim it.
+///
+/// The first two of these options may also be set directly on a
+/// [`Sending`] object, in which case they only apply to that operation.
 #[derive(Debug)]
-pub struct Sender<T> {
-    inner: Arc<Inner<T>>,
-    done: bool,
+pub struct Sender<'a, T> {
+    hatch: Option<Holder<'a, Hatch<T>>>,
+    flags: Flags,
 }
 
-impl<T> Sender<T> {
+impl<'a, T> Sender<'a, T> {
+    /// Creates a new Sender.
+    ///
+    /// # Safety
+    ///
+    /// You must not permit multiple live senders to exist.
     #[inline(always)]
-    pub(crate) fn new(inner: Arc<Inner<T>>) -> Self {
-        Sender { inner, done: false }
+    pub(crate) unsafe fn new(hatch: Holder<'a, Hatch<T>>) -> Self {
+        Sender { hatch: Some(hatch), flags: DEFAULT }
     }
 
-    /// Closes the channel by causing an immediate drop
+    /// Returns a disposable object for a single send operation.
     #[inline(always)]
-    pub fn close(self) { }
+    #[must_use = "Sending does nothing unless you call `.now` or poll it as a Future."]
+    pub fn send<'b>(&'b mut self, value: T) -> Sending<'a, 'b, T> {
+        let flags = self.flags;
+        Sending { sender: Some(self), value: Some(value), flags }
+    }
 
-    /// true if the channel is closed
-    #[inline(always)]
-    pub fn is_closed(&self) -> bool { self.inner.state().closed() }
-
-    /// Waits for a Receiver to be waiting for us to send something
-    /// (i.e. allows you to produce a value to send on demand).
-    /// Fails if the Receiver is dropped.
-    #[inline]
-    pub fn wait(self) -> impl Future<Output = Result<Self, Closed>> {
-        let mut this = Some(self);
-        poll_fn(move |ctx| {
-            let mut that = this.take().unwrap();
-            let state = that.inner.state();
-            if state.closed() {
-                that.done = true;
-                Poll::Ready(Err(Closed()))
-            } else if state.recv() {
-                Poll::Ready(Ok(that))
+    #[cfg(feature="async")]
+    #[must_use = "Wait does nothing unless you oll it as a Future."]
```

Comment:
I don't think this `must_use` is required, because if `Wait` is a `Future`, it would issue a warning anyways.
And the comment has a typo: `oll`.

Reply:
I don't think they're automatic. I guess I'll have a play and figure out. And fix the typo.